### ISSUE TITLE
When using codemeta.json files, we need to give guidance on race conditions

### ISSUE THREAD

**Issue by npch** [codemeta/codemeta#167]
If we consider the use case that codemeta.json files are used to represent metadata about a piece of scientific software, and that they reside in the root of the code repository, there are potential problems caused by the fact it is both describing the repository and part of the repository.

These include:
- what is the dateModified? Conceptually, this should be the date of the last commit on the source code, however in updating the codemeta.json, you are also modifying the repository.
- in generating a DOI for the software, if you cannot "pre-reserve" a DOI automatically, then the DOI included in the codemeta.json file will be out of sync.

Noted at FORCE2017 Hackathon.

**Comment by cboettig** [comment]
Definitely would be nice to have some ideas on how to proceed on this one.  Related to https://github.com/codemeta/codemeta/issues/24#issuecomment-164676080.  @lnielsen I don't suppose there's a workflow for pre-reserving DOIs that could be automated in the Zenodo import?

**Comment by mbjones** [comment]
Many repositories support pre-reserving DOIs and other identifiers.  That's what we do in the KNB, and it is supported via the EZID API for issuing DOIs using the `reserved` status.  It is commonplace that we need to embed an identifier for an object in that object or in its identifier before it is published to an archive.  As many archives already support this, it seems like Zenodo could as well, and its super useful.

**Comment by cboettig** [comment]
Yup, https://github.com/codemeta/codemeta/issues/24#issuecomment-164676080 Lars says Zenodo already supports the ability to reserve a DOI manually (though it's not obvious to me how to do it from either the API or web interface) and then upload to that.  It's just not supported through the automatic import tool, which presently just detects the creation of a release on GitHub.  

I think the automatic integration Zenodo uses with GitHub releases is pretty compelling and has obviously seen some significant adoption.  It would be nice if when "Flipping the switch" on a repo in zenodo (activating the watching for release) if Zenodo could automatically reserve a DOI and display it to the user, so it could be added to the `codemeta.json` (and as Badge on the README) prior to actually creating the GitHub release that will trigger the import into Zenodo.

Something similar on the DataONE/KNB would also be awesome...

**Comment by lnielsen** [comment]
We mint a DOI for every new release, hence the codemeta should be updated on every release (meaning we need commit access to the repo, and need to update the file before downloading the zip-ball). 

It's similar to software versioning. When you make a new release of software, you usually need to bump the version number in one or more files, then finally you tag the software and push it. For pre-registering DOIs I think this should work the same way. You bump the version number and bump the DOI in the codemeta (ideally in a single CLI tool call). On the Zenodo-end we would need a) a  CLI tool to pre-register the DOI for the given repository and b) make the Zenodo-GitHub integration aware of that pre-registered DOI.

Re. how to do it on Zenodo (just click the button):
![screen shot 2017-10-30 at 13 57 51](https://user-images.githubusercontent.com/1698163/32172014-6a3f45e0-bd7a-11e7-88cd-0d0f9c758bd9.png)

**Comment by cboettig** [comment]
@lnielsen Very nice; I should have realized I had to start a new upload to see the 'reserve doi' button.  It would be great to have this button appear when a user activates as GitHub repo in the Zenodo interface prior to creating a release, so that they can update the necessary metadata files in the usual order as you suggest without having to do a manual upload.  (Though I guess there will still be plenty of cases where a user forgets to update the DOI in the package metadata file before creating a new release; it's not clear to me how best to handle those inevitable conflicts).  

Is the "Reserve DOI" option available from the API?  That would facilitate a scripted upload locally that could automate the dance of reserving DOI, adding to metadata, etc; either uploading directly to Zenodo or by tagging as a release in GitHub.

**Comment by cboettig** [comment]
@lnielsen Also, I'm assuming there's no way to get the 'versioned" / linked DOIs for subsequent releases if users start reserving a DOI via manual upload interface?  It would be nice if it were possible to "reserve my next DOI" for any GitHub repo linked to zenodo

**Comment by nuest** [comment]
@cboettig You can reserve a DOI when you create a new deposition via the API, search "prereserve_doi" on http://developers.zenodo.org/#representation It would indeed be nice if the GitHub-Zenodo process would include a step to be able to edit files in GitHub with a pre-reserved DOI!

**Comment by lnielsen** [comment]
Here's an quick example:
```python
>>> headers = {"Content-Type": "application/json"}
>>> r = requests.post('https://zenodo.org/api/deposit/depositions',
...                   params={'access_token': MY_ACCESS_TOKEN}, json={},
...                   headers=headers)
>>> r.status_code
201
>>> r.json()
```

```json
{
  "created": "2016-06-15T16:10:03.319363+00:00",
  "files": [],
  "id": 1234,
  "links": {
    "discard": "https://zenodo.org/api/deposit/depositions/1234/actions/discard",
    "edit": "https://zenodo.org/api/deposit/depositions/1234/actions/edit",
    "files": "https://zenodo.org/api/deposit/depositions/1234/files",
    "publish": "https://zenodo.org/api/deposit/depositions/1234/actions/publish",
    "newversion": "https://zenodo.org/api/deposit/depositions/1234/actions/newversion",
    "self": "https://zenodo.org/api/deposit/depositions/1234"
  },
  "metadata": {
    "prereserve_doi": {
      "doi": "10.5072/zenodo.1234",
      "recid": 1234
    }
  },
  "modified": "2016-06-15T16:10:03.319371+00:00",
  "owner": 1,
  "record_id": 1234,
  "state": "unsubmitted",
  "submitted": false,
  "title": ""
}
```

As for the Zenodo-GitHub pre-register DOI option, then it's not straight-forward to integrate due to the chicken-egg problem. We would need to first create "an empty deposit" for each GitHub repo when you flip the switch and after each release, that would sit and wait for the next release on GitHub. This would allow you to see what would be the next DOI for your repository. This is not how it works today, and there would be some bookkeeping and cleanup tasks to implement to make this work. And again, I think this is actually something that should be integrated into your software release process with some automation so that you bump the version number and DOI  with a single CLI tool that does it all (otherwise I'm sure there will be all sorts of ways to forget to update the DOI, which will result in file update requests on or side :-) ).

**Comment by cboettig** [comment]
@lnielsen Thanks, yup, definitely a tricky problem.  I agree entirely that we want the steps automated so that a single command makes sure all the steps are done and done in the right order, otherwise bits will inevitably be missed.  

Thanks for the pointer on the API -- so basically creating a new deposit without any upload data merely pre-reserves the DOI?  Is there a way in the API to identify subsequent releases as associated with the previous release?  I think the way Zenodo now does versioning with DOIs (i.e. subsequent releases are linked in the Zenodo interface and there is a 'master' DOI to refer to all versions/latest version) is really helpful; and I think if I naively implemented a CLI to create releases by pre-reserving a DOI, then I would not be getting this connection, right?  Is there a way to "pre-reserve my versioned DOI?"

I'm not much of a web developer, but it's not obvious to me that this has to be done with a CLI client.  Could Zenodo use the GitHub API to handle the changes at that end (i.e. checking/updating the `codemeta.json` file?


Some users have suggested a far more trivial solution of simply using Zenodo's 'master DOI' that always points to the latest version as the DOI in their metadata files (`codemeta.json`, possibly elsewhere like a CITATION file in an R package).  No doubt this will be unpopular for repository types (deepening the issues already in dispute re having such a DOI in the first place), but I must give it points for being eminently simple and practical.  After the first release, no more need to worry about forgetting to update the DOI listed somewhere!

**Comment by mbjones** [comment]
The 'master DOI' approach has the advantage of being simple. We have the same feature in DataONE (we call it a 'Series Identifier' because it identifies a series of versions) because it is in widespread demand.  But, it has the disadvantage of preventing citation and usage tracking for specific versions/releases of software.  This goes directly against the [Force 11 Software Citation Principles guideline 6](https://www.force11.org/software-citation-principles):

> 6. Specificity: Software citations should facilitate identification of, and access to, the specific version of software that was used. Software identification should be as specific as necessary, such as using version numbers, revision numbers, or variants such as platforms.

As well as the [Force 11 Joint Declaration of Data Citation Principles guideline 7](https://www.force11.org/group/joint-declaration-data-citation-principles-final):

> 7. Specificity and Verifiability
Data citations should facilitate identification of, access to, and verification of the specific data that support a claim. Citations or citation metadata should include information about provenance and fixity sufficient to facilitate verfiying that the specific timeslice, version and/or granular portion of data retrieved subsequently is the same as was originally cited.

 Even though its in demand, and even though its simpler, I still think its worth the effort to build systems that allow people to cite the specific version used.
